### ISSUE TITLE
Infinitely-Stuck Process Issue

### ISSUE THREAD

**Issue by djspiewak** [scalaz/scalaz-stream#254]
Warning: I have yet to be able to minimize this issue.  When I run into it, the symptoms are extremely deterministic, but I haven't been able to pare it down to anything more narrow than the scope of a test for a relatively complex component of my application.  The only commonality thus far seems to be the use of a `Signal` and `wye.interrupt` to asynchronously control the termination of a process (which I commonly use in tests).  Even that isn't certain though.

Basically what is happening is scalaz-stream appears to be stuck in an trampolined, infinitely recursive computation.  The exact stack trace I get varies dramatically depending on when I do the thread dump, but it's usually something pretty close to the following:

```
"pool-2-thread-7" #18 daemon prio=5 os_prio=31 tid=0x00007f8d0c3ed800 nid=0x6303 runnable [0x0000000127574000]
   java.lang.Thread.State: RUNNABLE
	at scala.collection.immutable.VectorBuilder.$plus$eq(Vector.scala:720)
	at scala.collection.immutable.VectorBuilder.$plus$eq(Vector.scala:701)
	at scala.collection.generic.Growable$$anonfun$$plus$plus$eq$1.apply(Growable.scala:59)
	at scala.collection.generic.Growable$$anonfun$$plus$plus$eq$1.apply(Growable.scala:59)
	at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
	at scala.collection.mutable.WrappedArray.foreach(WrappedArray.scala:35)
	at scala.collection.generic.Growable$class.$plus$plus$eq(Growable.scala:59)
	at scala.collection.immutable.VectorBuilder.$plus$plus$eq(Vector.scala:725)
	at scala.collection.immutable.VectorBuilder.$plus$plus$eq(Vector.scala:701)
	at scala.collection.generic.GenericCompanion.apply(GenericCompanion.scala:49)
	at scalaz.  fromstream.Process$.emit(Process.scala:741)
	at scalaz.stream.Process$$anonfun$await1$1.apply(Process.scala:722)
	at scalaz.stream.Process$$anonfun$await1$1.apply(Process.scala:722)
	at scalaz.stream.Process$$anonfun$awaitOr$1$$anonfun$apply$32$$anonfun$apply$33$$anonfun$apply$35.apply(Process.scala:717)
	at scalaz.stream.Process$$anonfun$awaitOr$1$$anonfun$apply$32$$anonfun$apply$33$$anonfun$apply$35.apply(Process.scala:717)
	at scalaz.$bslash$div.fold(Either.scala:57)
	at scalaz.stream.Process$$anonfun$awaitOr$1$$anonfun$apply$32$$anonfun$apply$33.apply(Process.scala:717)
	at scalaz.stream.Process$$anonfun$awaitOr$1$$anonfun$apply$32$$anonfun$apply$33.apply(Process.scala:717)
	at scalaz.stream.Util$.Try(Util.scala:42)
	at scalaz.stream.Process$$anonfun$awaitOr$1$$anonfun$apply$32.apply(Process.scala:717)
	at scalaz.stream.Process$$anonfun$awaitOr$1$$anonfun$apply$32.apply(Process.scala:717)
	at scalaz.Trampoline$$anonfun$delay$1.apply(Free.scala:234)
	at scalaz.Trampoline$$anonfun$delay$1.appl space 128K,  99% usedy(Free.scala:234)
	at scalaz.std.FunctionInstances$$anon$1$$anonfun$map$1.apply(Function.scala:56)
	at scalaz.Free$$anonfun$run$1.apply(Free.scala:172)
	at scalaz.Free$$anonfun$run$1.apply(Free.scala:172)
	at scalaz.Free.go2$1(Free.scala:119)
	at scalaz.Free.go(Free.scala:122)
	at scalaz.Free.run(Free.scala:172)
	at scalaz.stream.process1$Await1$$anonfun$unapply$1$$anonfun$apply$31.apply(process1.scala:622)
	at scalaz.stream.process1$Await1$$anonfun$unapply$1$$anonfun$apply$31.apply(process1.scala:622)
	at scalaz.stream.Util$.Try(Util.scala:42)
	at scalaz.stream.process1$Await1$$anonfun$unapply$1.apply(process1.scala:622)
	at scalaz.stream.process1$Await1$$anonfun$unapply$1.apply(process1.scala:622)
	at scalaz.stream.process1$.go$5(process1.scala:153)
	at scalaz.stream.process1$.feed(process1.scala:160)
	at scalaz.stream.Process$$anonfun$pipe$1.apply(Process.scala:137)
	at scalaz.stream.Process$$anonfun$pipe$1.apply(Process.scala:131)
	at scalaz.stream.Process$$anonfun$flatMap$1.apply(Process.scala:40)
	at scalaz.stream.Process$$anonfun$flatMap$1.apply(Process.scala:40)
	at scalaz.stream.Util$.Try(Util.scala:42)
	at scalaz.stream.Process$class.flatMap(Process.scala:40)
	at scalaz.stream.Process$Emit.flatMap(Process.scala:572)
	at scalaz.stream.Process$$anonfun$flatMap$4.apply(Process.scala:42)
	at scalaz.stream.Process$$anonfun$flatMap$4.apply(Process.scala:42)
	at scalaz.Free$$anonfun$map$1.apply(Free.scala:52)
	at scalaz.Free$$anonfun$map$1.apply(Free.scala:52)
	at scalaz.Free.resume(Free.scala:73)
	at scalaz.Free.go2$1(Free.scala:118)
	at scalaz.Free.go(Free.scala:122)
	at scalaz.Free.run(Free.scala:172)
	at scalaz.stream.Process$$anonfun$go$1$1.apply(Process.scala:75)
	at scalaz.stream.Process$$anonfun$go$1$1.apply(Process.scala:75)
	at scalaz.stream.Util$.Try(Util.scala:42)
	at scalaz.stream.Process$class.go$1(Process.scala:75)
	at scalaz.stream.Process$class.step(Process.scala:88)
	at scalaz.stream.Process$Append.step(Process.scala:612)
	at scalaz.stream.Process$$anonfun$pipe$1.apply(Process.scala:135)
	at scalaz.stream.Process$$anonfun$pipe$1.apply(Process.scala:131)
	at scalaz.stream.Process$$anonfun$flatMap$1.apply(Process.scala:40)
	at scalaz.stream.Process$$anonfun$flatMap$1.apply(Process.scala:40)
	at scalaz.stream.Util$.Try(Util.scala:42)
	at scalaz.stream.Process$class.flatMap(Process.scala:40)
	at scalaz.stream.Process$Emit.flatMap(Process.scala:572)
	at scalaz.stream.Process$$anonfun$flatMap$4.apply(Process.scala:42)
	at scalaz.stream.Process$$anonfun$flatMap$4.apply(Process.scala:42)
	at scalaz.Free$$anonfun$map$1.apply(Free.scala:52)
	at scalaz.Free$$anonfun$map$1.apply(Free.scala:52)
	at scalaz.Free$$anonfun$flatMap$1$$anonfun$apply$1.apply(Free.scala:60)
	at scalaz.Free$$anonfun$flatMap$1$$anonfun$apply$1.apply(Free.scala:60)
	at scalaz.Free.resume(Free.scala:72)
	at scalaz.Free.go2$1(Free.scala:118)
	at scalaz.Free.go(Free.scala:122)
	at scalaz.Free.run(Free.scala:172)
	at scalaz.stream.Process$$anonfun$go$1$1.apply(Process.scala:75)
	at scalaz.stream.Process$$anonfun$go$1$1.apply(Process.scala:75)
	at scalaz.stream.Util$.Try(Util.scala:42)
	at scalaz.stream.Process$class.go$1(Process.scala:75)
	at scalaz.stream.Process$class.step(Process.scala:88)
	at scalaz.stream.Process$Append.step(Process.scala:612)
	at scalaz.stream.Process$$anonfun$tee$1.apply(Process.scala:175)
	at scalaz.stream.Process$$anonfun$tee$1.apply(Process.scala:173)
	at scalaz.stream.Process$$anonfun$flatMap$1.apply(Process.scala:40)
	at scalaz.stream.Process$$anonfun$flatMap$1.apply(Process.scala:40)
	at scalaz.stream.Util$.Try(Util.scala:42)
	at scalaz.stream.Process$class.flatMap(Process.scala:40)
	at scalaz.stream.Process$Emit.flatMap(Process.scala:572)
	at scalaz.stream.Process$$anonfun$flatMap$4.apply(Process.scala:42)
	at scalaz.stream.Process$$anonfun$flatMap$4.apply(Process.scala:42)
	at scalaz.Free$$anonfun$map$1.apply(Free.scala:52)
	at scalaz.Free$$anonfun$map$1.apply(Free.scala:52)
	at scalaz.Free$$anonfun$flatMap$1$$anonfun$apply$1.apply(Free.scala:60)
	at scalaz.Free$$anonfun$flatMap$1$$anonfun$apply$1.apply(Free.scala:60)
	at scalaz.Free.resume(Free.scala:72)
	at scalaz.Free.go2$1(Free.scala:118)
	at scalaz.Free.go(Free.scala:122)
	at scalaz.Free.run(Free.scala:172)
	at scalaz.stream.Process$$anonfun$go$1$1.apply(Process.scala:75)
	at scalaz.stream.Process$$anonfun$go$1$1.apply(Process.scala:75)
	at scalaz.stream.Util$.Try(Util.scala:42)
	at scalaz.stream.Process$class.go$1(Process.scala:75)
	at scalaz.stream.Process$class.step(Process.scala:88)
	at scalaz.stream.Process$Append.step(Process.scala:612)
	at scalaz.stream.Process$$anonfun$pipe$1.apply(Process.scala:135)
	at scalaz.stream.Process$$anonfun$pipe$1.apply(Process.scala:131)
	at scalaz.stream.Process$$anonfun$flatMap$1.apply(Process.scala:40)
	at scalaz.stream.Process$$anonfun$flatMap$1.apply(Process.scala:40)
	at scalaz.stream.Util$.Try(Util.scala:42)
	at scalaz.stream.Process$class.flatMap(Process.scala:40)
	at scalaz.stream.Process$Emit.flatMap(Process.scala:572)
	at scalaz.stream.Process$$anonfun$flatMap$4.apply(Process.scala:42)
	at scalaz.stream.Process$$anonfun$flatMap$4.apply(Process.scala:42)
	at scalaz.Free$$anonfun$map$1.apply(Free.scala:52)
	at scalaz.Free$$anonfun$map$1.apply(Free.scala:52)
	at scalaz.Free$$anonfun$flatMap$1$$anonfun$apply$1.apply(Free.scala:60)
	at scalaz.Free$$anonfun$flatMap$1$$anonfun$apply$1.apply(Free.scala:60)
	at scalaz.Free.resume(Free.scala:72)
	at scalaz.Free.go2$1(Free.scala:118)
	at scalaz.Free.go(Free.scala:122)
	at scalaz.Free.run(Free.scala:172)
	at scalaz.stream.Process$$anonfun$go$1$1.apply(Process.scala:75)
	at scalaz.stream.Process$$anonfun$go$1$1.apply(Process.scala:75)
	at scalaz.stream.Util$.Try(Util.scala:42)
	at scalaz.stream.Process$class.go$1(Process.scala:75)
	at scalaz.stream.Process$class.step(Process.scala:88)
	at scalaz.stream.Process$Append.step(Process.scala:612)
	at scalaz.stream.Process$SourceSyntax.scalaz$stream$Process$SourceSyntax$$runStep$1(Process.scala:1330)
	at scalaz.stream.Process$SourceSyntax.runAsync(Process.scala:1381)
	at scalaz.stream.wye$$anonfun$apply$42.scalaz$stream$wye$$anonfun$$runSide$1(wye.scala:680)
	at scalaz.stream.wye$$anonfun$apply$42$$anonfun$4.apply(wye.scala: total 523008K, used 216799K684)
	at scalaz.stream.wye$$anonfun$apply$42$$anonfun$4.apply(wye.scala:684)
	at scalaz.stream.wye$$anonfun$apply$42.go$12(wye.scala:801)
	at scalaz.stream.wye$$anonfun$apply$42.scalaz$stream$wye$$anonfun$$runY$1(wye.scala:813)
	at scalaz.stream.wye$$anonfun$apply$42$$anonfun$apply$43.apply(wye.scala:834)
	at scalaz.stream.wye$$anonfun$apply$42$$anonfun$apply$43.apply(wye.scala:818)
	at scalaz.concurrent.Actor.scalaz$concurrent$Actor$$act(Actor.scala:50)
	at scalaz.concurrent.Actor$$anonfun$schedule$1.apply$mcV$sp(Actor.scala:46)
	at scalaz.concurrent.Actor$$anonfun$schedule$1.apply(Actor.scala:46)
	at scalaz.concurrent.Actor$$anonfun$schedule$1.apply(Actor.scala:46)
	at scalaz.concurrent.StrategysLow$$anon$3$$anon$4.call(Strategy.scala:70)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:744)
```

In one particularly memorable incident, I was driving a scalacheck property where the implementation of the property exhibited this issue on certain inputs.  Within a few attempts, all eight of my fixed thread pool workers looked like the above, starving out all subsequent computations.  Oh, and since the timeouts in the tests were implemented using `wye.interrupt` (thus requiring at least *a* thread on which to do work), the end result was a lot of processor time and no progress.  Hurray!

And yes, it will remain in this state indefinitely.  Or at least, as close to indefinitely as I cared to let it run (I've let it go as long as an hour).

I'm trying to minimize this whenever I get the chance, but it's rough going.  I'll update this issue if I see a way to narrow down the cause a bit more.

**Comment by pchlupacek** [comment]
@djspiewak can you please verify if proposed #271 fix fixes this issue?

**Comment by djspiewak** [comment]
Actually the stack trace listed in the OP looks suspiciously like the `step` issue that @alissapajer and I have been chasing.

**Comment by pchlupacek** [comment]
Well we will see. Anyhow I have posted fix to #271 so would be excellent if you can test this with that branch too..

**Comment by djspiewak** [comment]
Cleaning this up.  Let's just throw this under the umbrella of all the other "stuck process" issues we've run into.
